File: app/src/main/java/isden/mois/magellanlauncher/Onyx.java
package isden.mois.magellanlauncher;

import android.content.Context;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.support.v4.util.Pair;

import isden.mois.magellanlauncher.helpers.DBBooks;
import isden.mois.magellanlauncher.holders.BookTime;
import isden.mois.magellanlauncher.holders.HistoryDetail;
import isden.mois.magellanlauncher.utils.DateKt;

import java.io.File;
import java.util.*;

public class Onyx {
    private static final String CONTENT_URI = "content://com.onyx.android.sdk.OnyxCmsProvider/";

    public static Metadata getCurrentBook(Context ctx) {
        Cursor c = ctx.getContentResolver().query(
                Uri.parse(CONTENT_URI + "current_book"),
                null,
                null,
                null,
                null
        );

        return parseBook(ctx, c);
    }

    public static Metadata getBook(Context ctx, String MD5) {
        Cursor c = ctx.getContentResolver().query(
                Uri.parse(CONTENT_URI + "book"),
                null,
                null,
                new String[]{MD5},
                null
        );

        return parseBook(ctx, c);
    }

    private static Metadata parseBook(Context ctx, Cursor c) {
        Metadata metadata = null;

        if (c != null) {
            if (c.moveToFirst()) {
                metadata = new Metadata();
                metadata.md5 = c.getString(c.getColumnIndex("MD5"));
                metadata.title = c.getString(c.getColumnIndex("Title"));
                metadata.author = c.getString(c.getColumnIndex("Authors"));
                metadata.thumbnail = c.getString(c.getColumnIndex("Thumbnail"));
                metadata.filePath = c.getString(c.getColumnIndex("Location"));

                String progress = c.getString(c.getColumnIndex("Progress"));
                if (progress != null && progress.contains("/")) {
                    String[] progressData = progress.split("/");
                    metadata.progress = Integer.parseInt(progressData[0]);
                    metadata.size = Integer.parseInt(progressData[1]);
                }
            }
            c.close();
        }

        if (metadata != null) {
            Pair<BookTime, Integer> details = getBookDetails(ctx, metadata.md5);
            metadata.time = details.first;
            metadata.readPages = details.second;
        }

        return metadata;
    }

    public static Pair<BookTime, Integer> getBookDetails(Context ctx, String MD5) {
        ArrayList<int[]> speeds = new ArrayList<>();
        int lastProgress = 0;
        BookTime bookTime = new BookTime();
        bookTime.totalTime = 0;
        bookTime.currentTime = 0;
        String[] progressData = null;

        Cursor c = getHistoryCursor(ctx, MD5);
        if (c != null) {
            try {
                while (c.moveToNext()) {
                    long time = c.getLong(c.getColumnIndex("Time"));
                    String progress = c.getString(c.getColumnIndex("Progress"));
                    if (progress.contains("/")) {
                        progressData = progress.split("/");
                        int currentProgress = Integer.parseInt(progressData[0]);

                        // Если начал читать заново.
                        if (currentProgress < lastProgress) {
                            int totalProgress = Integer.parseInt(progressData[1]);
                            int currentPercent = 100 * currentProgress / totalProgress;
                            int previousPercent = 100 * lastProgress / totalProgress;
                            int diffPercent = previousPercent - currentPercent;

                            if (currentPercent < 20 && diffPercent > 20) {
                                bookTime.currentTime = 0;
                                speeds = new ArrayList<>();
                                lastProgress = 0;
                            }
                        }

                        bookTime.totalTime += time;
                        bookTime.currentTime += time;
                        int speed;

                        if (currentProgress != lastProgress) {
                            speed = (int) (time / (currentProgress - lastProgress));
                        } else {
                            speed = 0;
                        }

                        // Если были пропущены главы.
                        if (currentProgress > lastProgress && speed > 10000 && speed < 100000) {
                            speeds.add(new int[]{speed, currentProgress - lastProgress});
                        }

                        lastProgress = currentProgress;
                    }
                }
            } finally {
                c.close();
            }
        }

        long bookWeight = 0;
        double bookPages = 0;
        for (int[] speedTime : speeds) {
            bookWeight += (long) speedTime[0] * speedTime[1];
            bookPages += speedTime[1];
        }

        bookTime.speed = bookWeight / bookPages;

        return new Pair<>(bookTime, (int) Math.round(bookPages));
    }


    public static String getTotalTime(Context ctx) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(ctx);
        String limit = prefs.getString("history_clean_limit", "20000");

        Cursor c = ctx.getContentResolver().query(
                Uri.parse(CONTENT_URI + "library_history"),
                new String[]{"SUM(EndTime - StartTime) AS Time"},
                "(EndTime - StartTime) > " + limit,
                null,
                null
        );

        if (c != null) {
            if (c.moveToFirst()) {
                long totalTime = c.getLong(0);
                return DateKt.formatHumanTime(totalTime);
            }
            c.close();
        }
        return "0";
    }

    public static void cleanDirtyHistory(Context ctx) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(ctx);
        String limit = prefs.getString("history_clean_limit", "20000");

        ctx.getContentResolver().delete(
                Uri.parse(CONTENT_URI + "library_history"),
                "(EndTime - StartTime) < " + limit,
                null
        );
    }

    public static List<Metadata> getLastDownloaded(Context ctx, int limit) {
        Cursor c = ctx.getContentResolver().query(
                Uri.parse(CONTENT_URI + "last_downloaded"),
                null,
                null,
                null,
                String.valueOf(limit)
        );

        List<Metadata> metadataList = new ArrayList<>();

        if (c != null) {
            try {
                while (c.moveToNext()) {
                    Metadata metadata = new Metadata();
                    metadata.author = c.getString(c.getColumnIndex("Authors"));
                    metadata.title = c.getString(c.getColumnIndex("Title"));
                    metadata.filePath = c.getString(c.getColumnIndex("Location"));
                    metadata.thumbnail = c.getString(c.getColumnIndex("Thumbnail"));

                    metadataList.add(metadata);
                }
            } finally {
                c.close();
            }
        }

        return metadataList;
    }

    public static Cursor getHistoryCursor(Context ctx, String MD5) {
        if (MD5 == null) {
            return null;
        }

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(ctx);
        String limit = prefs.getString("history_clean_limit", "20000");

        return ctx.getContentResolver().query(
                Uri.parse(CONTENT_URI + "library_history"),
                new String[]{"StartTime", "(EndTime - StartTime) AS Time", "Progress"},
                "Progress <> \"5/5\" AND MD5 = ? AND (EndTime - StartTime) > " + limit,
                new String[]{MD5},
                "StartTime"
        );
    }

}

File: app/src/main/java/isden/mois/magellanlauncher/utils/Onyx.kt
package isden.mois.magellanlauncher.utils

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.preference.PreferenceManager
import android.util.Log
import isden.mois.magellanlauncher.helpers.DBBooks
import isden.mois.magellanlauncher.models.BookMetadata
import isden.mois.magellanlauncher.models.BookTime
import isden.mois.magellanlauncher.models.HistoryDetail
import java.util.*
import kotlin.collections.ArrayList

/**
 * Created by isden on 16.06.17.
 */

fun getRecentReading(ctx: Context, limit: Int = 0): List<BookMetadata> {
    val result = LinkedList<BookMetadata>()

    val builder = StringBuilder("SELECT ")
    builder.append("m.MD5, Authors, Title, Name, NativeAbsolutePath, m.Progress, ")
    builder.append("MAX(LastAccess) AS LastAccess, SUM(h.EndTime - h.StartTime) AS TotalTime,")
    builder.append("t._data as Thumbnail, MIN(h.StartTime) AS FirstTime ")
    builder.append("FROM library_metadata m ")
    builder.append("LEFT JOIN library_history h ON h.MD5 = m.MD5 ")
    builder.append("LEFT JOIN library_thumbnail t ON Source_MD5 = m.MD5 AND Thumbnail_Kind = \"Middle\" ")
    builder.append("WHERE Title IS NOT NULL AND Name LIKE \"%.fb2%\" ")
    builder.append("GROUP BY m.MD5 ")
    builder.append("ORDER BY LastAccess DESC")
    if (limit > 0) {
        builder.append("LIMIT ")
        builder.append(limit)
    }

    val query = builder.toString()
    val dbBooks = DBBooks(ctx)
    val db = dbBooks.readableDatabase

    try {
        val c = db.rawQuery(query, emptyArray<String>())
        if (c != null) {
            if (c.moveToFirst()) {
                do {
                    result.add(createMetadata(c))
                } while (c.moveToNext())
            }
            c.close()
        }
    } finally {
        db.close()
    }

    return result
}

fun getBooks(ctx: Context, status: String, order: String): List<BookMetadata> {
    val result = ArrayList<BookMetadata>()
    val query = "SELECT m.MD5, Authors, Title, Name, NativeAbsolutePath, m.Progress, " +
            "MAX(LastAccess) AS LastAccess, SUM(h.EndTime - h.StartTime) AS TotalTime, " +
            "t._data as Thumbnail, MIN(h.StartTime) AS FirstTime " +
            "FROM library_metadata m " +
            "LEFT JOIN library_history h ON h.MD5 = m.MD5 " +
            "LEFT JOIN library_thumbnail t ON Source_MD5 = m.MD5 AND Thumbnail_Kind = \"Middle\" " +
            "WHERE Title IS NOT NULL AND (Name LIKE \"%.fb2\" OR Name LIKE \"%.epub\") " +
            "AND ${status} " +
            "GROUP BY m.MD5 " +
            "ORDER BY ${order} DESC"
    val dbBooks = DBBooks(ctx)
    val db = dbBooks.readableDatabase

    try {
        val c = db.rawQuery(query, emptyArray<String>())
        if (c != null) {
            if (c.moveToFirst()) {
                do {
                    result.add(createMetadata(c))
                } while (c.moveToNext())
            }
            c.close()
        }
    } finally {
        db.close()
    }

    return result
}

fun getDetailedHistory(ctx: Context, md5: String): Array<HistoryDetail> {
    var historyDetails = emptyArray<HistoryDetail>()
    val prefs = PreferenceManager.getDefaultSharedPreferences(ctx)
    val limit = prefs.getString("history_clean_limit", "20000").toInt()

    val builder = StringBuilder("SELECT ")
    builder.append("StartTime, (EndTime - StartTime) AS Time, Progress ")
    builder.append("FROM library_history ")
    builder.append("WHERE Progress <> \"5/5\" AND MD5 = ? ")
    builder.append("ORDER BY EndTime")

    val query = builder.toString()
    val dbBooks = DBBooks(ctx)
    val db = dbBooks.readableDatabase

    try {
        val c = db.rawQuery(query, arrayOf(md5))
        if (c != null) {
            if (c.moveToFirst()) {
                val dates = Hashtable<String, HistoryDetail>();
                var lastProgress = 0;

                do {
                    val startTime = c.getLong(c.getColumnIndex("StartTime"));
                    val readTime = c.getLong(c.getColumnIndex("Time"));
                    val progressStr = c.getString(c.getColumnIndex("Progress"))
                    val date = formatDate(startTime);
                    val progress = Integer.parseInt(progressStr.substring(0, progressStr.indexOf('/')))
                    val pages = progress - lastProgress
                    val speed = if (pages > 0) readTime / pages else 0

                    Log.d("HISTORY", "readTime: " + readTime + "; lastProgress: " + lastProgress + "; progress: " + progress + "; pages: " + pages + "; speed: " + speed)

                    if (readTime > limit && progress > lastProgress && speed > 15000 && speed < 100000) {
                        if (dates.containsKey(date)) {
                            val detail = dates[date];
                            detail!!.progress = progress;
                            detail.spent += readTime;
                            detail.pages += pages
                        } else {
                            val detail = HistoryDetail(date, startTime, readTime, progress, pages, 0);
                            dates[date] = detail
                        }
                    }

                    lastProgress = progress
                } while (c.moveToNext())
                historyDetails += dates.values
            }
        }
    } finally {
        db.close()
    }

    for (detail in historyDetails) {
        detail.speed = if (detail.pages > 0) Math.round((detail.spent / detail.pages).toDouble()).toInt() else 0
    }

    Arrays.sort(historyDetails, Comparator<HistoryDetail> { h1, h2 ->
        if (h1 == null && h2 == null) {
            return@Comparator 0
        }
        if (h1 == null) {
            return@Comparator 1
        }
        if (h2 == null) {
            return@Comparator -1
        }

        val t1 = h1.timestamp
        val t2 = h2.timestamp
        if (t1 < t2) -1 else if (t1 == t2) 0 else 1
    })

    return historyDetails
}

fun createMetadata(c: Cursor): BookMetadata {
    val md5 = c.getString(c.getColumnIndex("MD5"))
    val filename = c.getString(c.getColumnIndex("Name"))
    val title = c.getString(c.getColumnIndex("Title"))
    val author = c.getString(c.getColumnIndex("Authors"))
    val filePath = c.getString(c.getColumnIndex("NativeAbsolutePath"))
    val progress = c.getString(c.getColumnIndex("Progress"))
    val lastAccess = c.getLong(c.getColumnIndex("LastAccess"))
    val thumbnail = c.getString(c.getColumnIndex("Thumbnail"))
    val currentTime = c.getLong(c.getColumnIndex("TotalTime"))
    val firstTime = c.getLong(c.getColumnIndex("FirstTime"))

    val time = BookTime()
    time.currentTime = currentTime
    time.totalTime = time.totalTime

    return BookMetadata(md5, author, title, filename, filePath, lastAccess, time, thumbnail, progress, firstTime)
}

fun changeStatus(ctx: Context, book: BookMetadata, status: Int) {
    val dbBooks = DBBooks(ctx)
    val db = dbBooks.readableDatabase

    try {
        val values = ContentValues()
        values.put("Status", status)

        db.update("library_metadata", values, "MD5 = ?", arrayOf(book.md5))
    } finally {
        db.close()
    }
}
